File: E3_shiny_app/R/seed_catalogue.R
#' Known-E3 seed catalogue reconstructed from the loaded resource.

seed_catalogue_authority_relation <- function() "known_e3_seeds"

seed_catalogue_summary_relation_preference <- function() {
  c("candidate_evidence", "e3_cluster_candidate_evidence")
}

seed_catalogue_id_columns <- function() {
  c(
    "matched_seed_ids_calculated", "discovery_matched_seed_ids_calculated",
    "known_e3_seed_ids", "matched_e3_seeds"
  )
}

seed_catalogue_cluster_columns <- function() c("cluster_id", "representative_id")

seed_catalogue_annotation_fields <- function() {
  list(
    associated_seed_protein_names = c(
      "seed_protein_names", "discovery_seed_protein_names"
    ),
    associated_seed_categories = c(
      "seed_categories", "discovery_seed_categories"
    ),
    associated_seed_review_statuses = c(
      "seed_review_statuses", "discovery_seed_review_statuses"
    ),
    associated_seed_ubiquitin_go_statuses = c(
      "seed_ubiquitin_go_statuses", "discovery_seed_ubiquitin_go_statuses"
    ),
    associated_seed_organisms = c(
      "seed_organisms", "discovery_seed_organisms"
    )
  )
}

seed_catalogue_exact_annotation_fields <- function() {
  c(
    "seed_protein_names", "seed_category", "seed_review_status",
    "seed_ubiquitin_go_status", "seed_exclusion_go_term", "seed_organism",
    "seed_taxon_id", "seed_sequence_md5", "seed_evidence_type", "seed_source"
  )
}

seed_catalogue_first_column <- function(available, choices) {
  selected <- choices[choices %in% available]
  if (length(selected) == 0L) return(NULL)
  selected[[1L]]
}

#' Resolve seed evidence and optional sequence sources.
#'
#' @param relation_columns Named source columns by relation.
#' @return Capability metadata.
seed_catalogue_capability <- function(relation_columns) {
  summary_relation <- NULL
  summary_columns <- character()
  summary_seed_column <- NULL
  summary_cluster_column <- NULL
  for (relation in seed_catalogue_summary_relation_preference()) {
    columns <- relation_columns[[relation]]
    if (is.null(columns)) next
    seed_column <- seed_catalogue_first_column(
      available = columns,
      choices = seed_catalogue_id_columns()
    )
    if (is.null(seed_column)) next
    cluster_column <- seed_catalogue_first_column(
      available = columns,
      choices = seed_catalogue_cluster_columns()
    )
    summary_relation <- relation
    summary_columns <- columns
    summary_seed_column <- seed_column
    summary_cluster_column <- cluster_column
    break
  }
  sequence_columns <- relation_columns[["candidate_group_member_sequences"]]
  sequence_columns <- sequence_columns %||% character()
  authority <- seed_catalogue_authority_relation()
  authority_columns <- relation_columns[[authority]] %||% character()
  if ("seed_id" %in% authority_columns) {
    return(list(
      available = TRUE,
      mode = "authority",
      relation = authority,
      columns = authority_columns,
      seed_id_column = "seed_id",
      cluster_column = NULL,
      summary_relation = summary_relation,
      summary_columns = summary_columns,
      summary_seed_column = summary_seed_column,
      summary_cluster_column = summary_cluster_column,
      sequence_available = all(
        c("raw_identifier", "protein_sequence") %in% sequence_columns
      ),
      sequence_columns = sequence_columns
    ))
  }
  if (!is.null(summary_relation)) {
    return(list(
      available = TRUE,
      mode = "cluster_summary",
      relation = summary_relation,
      columns = summary_columns,
      seed_id_column = summary_seed_column,
      cluster_column = summary_cluster_column,
      summary_relation = summary_relation,
      summary_columns = summary_columns,
      summary_seed_column = summary_seed_column,
      summary_cluster_column = summary_cluster_column,
      sequence_available = all(
        c("raw_identifier", "protein_sequence") %in% sequence_columns
      ),
      sequence_columns = sequence_columns
    ))
  }
  list(
    available = FALSE,
    mode = NULL,
    relation = NULL,
    columns = character(),
    seed_id_column = NULL,
    cluster_column = NULL,
    summary_relation = NULL,
    summary_columns = character(),
    summary_seed_column = NULL,
    summary_cluster_column = NULL,
    sequence_available = FALSE,
    sequence_columns = character()
  )
}

seed_catalogue_aggregate <- function(available, choices, alias) {
  column <- seed_catalogue_first_column(available = available, choices = choices)
  if (is.null(column)) {
    return(paste0("CAST('' AS VARCHAR) AS ", quote_duckdb_identifier(alias)))
  }
  expression <- paste0(
    "nullif(trim(CAST(", quote_duckdb_identifier(column), " AS VARCHAR)), '')"
  )
  paste0(
    "coalesce(string_agg(DISTINCT ", expression, ", ';' ORDER BY ", expression,
    ") FILTER (WHERE ", expression, " IS NOT NULL), '') AS ",
    quote_duckdb_identifier(alias)
  )
}

seed_catalogue_sequence_cte <- function(capability, alias = "e3_resource") {
  if (!isTRUE(capability$sequence_available)) {
    return(paste0(
      "seed_sequences AS (SELECT CAST(NULL AS VARCHAR) AS seed_id, ",
      "CAST(NULL AS BIGINT) AS sequence_match_count, CAST(NULL AS BIGINT) AS ",
      "distinct_sequence_count, CAST(NULL AS VARCHAR) AS sequence_species, ",
      "CAST(NULL AS VARCHAR) AS sequence_identifiers, CAST(NULL AS VARCHAR) AS ",
      "protein_sequence WHERE FALSE)"
    ))
  }
  columns <- capability$sequence_columns
  accession <- if ("parsed_accession" %in% columns) {
    "CAST(parsed_accession AS VARCHAR)"
  } else {
    "CAST(NULL AS VARCHAR)"
  }
  species <- if ("species" %in% columns) {
    "CAST(species AS VARCHAR)"
  } else {
    "CAST('' AS VARCHAR)"
  }
  relation <- qualified_resource_relation(
    relation = "candidate_group_member_sequences",
    alias = alias
  )
  paste0(
    "sequence_rows AS (SELECT coalesce(nullif(trim(", accession, "), ''), ",
    "nullif(trim(CAST(raw_identifier AS VARCHAR)), '')) AS seed_id, ", species,
    " AS species, CAST(raw_identifier AS VARCHAR) AS raw_identifier, ",
    "nullif(trim(CAST(protein_sequence AS VARCHAR)), '') AS protein_sequence ",
    "FROM ", relation, "), seed_sequences AS (SELECT seed_id, ",
    "count(*) AS sequence_match_count, count(DISTINCT protein_sequence) FILTER ",
    "(WHERE protein_sequence IS NOT NULL) AS distinct_sequence_count, coalesce(",
    "string_agg(DISTINCT species, ';' ORDER BY species) FILTER (WHERE ",
    "nullif(trim(species), '') IS NOT NULL), '') AS sequence_species, coalesce(",
    "string_agg(DISTINCT raw_identifier, ';' ORDER BY raw_identifier) FILTER ",
    "(WHERE nullif(trim(raw_identifier), '') IS NOT NULL), '') AS ",
    "sequence_identifiers, min(protein_sequence) FILTER (WHERE protein_sequence ",
    "IS NOT NULL) AS protein_sequence FROM sequence_rows WHERE seed_id IS NOT ",
    "NULL GROUP BY seed_id)"
  )
}

seed_catalogue_json_text <- function(keys, available) {
  if (!"seed_metadata_json" %in% available) return("CAST('' AS VARCHAR)")
  candidates <- vapply(keys, function(key) {
    paste0(
      "nullif(trim(json_extract_string(TRY_CAST(seed_metadata_json AS JSON), '$.",
      key,
      "')), '')"
    )
  }, character(1L))
  paste0("coalesce(", paste(candidates, collapse = ", "), ", '')")
}

seed_catalogue_authority_source <- function(column, available) {
  if (!column %in% available) return("CAST('' AS VARCHAR)")
  paste0(
    "coalesce(CAST(", quote_duckdb_identifier(column), " AS VARCHAR), '')"
  )
}

seed_catalogue_cluster_summary_cte <- function(
  capability,
  alias = "e3_resource"
) {
  relation <- capability$summary_relation
  seed_column <- capability$summary_seed_column
  cluster_column <- capability$summary_cluster_column
  if (is.null(relation) || is.null(seed_column) || is.null(cluster_column)) {
    return(paste0(
      "cluster_summary AS (SELECT CAST(NULL AS VARCHAR) AS seed_id, ",
      "CAST(NULL AS BIGINT) AS source_cluster_count, CAST(NULL AS VARCHAR) AS ",
      "source_cluster_ids WHERE FALSE)"
    ))
  }
  source <- qualified_resource_relation(relation = relation, alias = alias)
  paste0(
    "cluster_links AS (SELECT DISTINCT trim(seed_id) AS seed_id, CAST(",
    quote_duckdb_identifier(cluster_column), " AS VARCHAR) AS source_cluster_id ",
    "FROM ", source, ", UNNEST(string_split(coalesce(CAST(",
    quote_duckdb_identifier(seed_column), " AS VARCHAR), ''), ';')) AS ",
    "seeds(seed_id) WHERE trim(seed_id) != ''), cluster_summary AS (SELECT ",
    "seed_id, count(DISTINCT nullif(trim(source_cluster_id), '')) AS ",
    "source_cluster_count, coalesce(string_agg(DISTINCT source_cluster_id, ';' ",
    "ORDER BY source_cluster_id) FILTER (WHERE nullif(trim(source_cluster_id), ",
    "'') IS NOT NULL), '') AS source_cluster_ids FROM cluster_links GROUP BY ",
    "seed_id)"
  )
}

build_seed_catalogue_authority_query <- function(
  capability,
  max_rows,
  alias = "e3_resource"
) {
  metadata <- list(
    seed_protein_names = c("protein_names", "protein_name", "name"),
    seed_category = c("category", "e3_category"),
    seed_review_status = c("reviewed", "review_status"),
    seed_ubiquitin_go_status = "ubiquitin_go_term",
    seed_exclusion_go_term = "exclusion_go_term",
    seed_organism = "organism",
    seed_taxon_id = c("organism_id", "taxon_id"),
    seed_sequence_md5 = "sequence_md5",
    seed_evidence_type = "evidence_type",
    seed_source = "source"
  )
  annotations <- vapply(names(metadata), function(output) {
    paste0(
      seed_catalogue_json_text(
        keys = metadata[[output]],
        available = capability$columns
      ),
      " AS ", quote_duckdb_identifier(output)
    )
  }, character(1L))
  provenance_columns <- c(
    "source_value", "source_column", "source_row", "source_path",
    "seed_metadata_json"
  )
  provenance <- vapply(provenance_columns, function(column) {
    paste0(
      seed_catalogue_authority_source(
        column = column,
        available = capability$columns
      ),
      " AS ", quote_duckdb_identifier(column)
    )
  }, character(1L))
  exact_projection <- paste0(
    "authority.",
    vapply(
      seed_catalogue_exact_annotation_fields(),
      quote_duckdb_identifier,
      character(1L)
    ),
    collapse = ", "
  )
  associated_blanks <- vapply(
    names(seed_catalogue_annotation_fields()),
    function(column) paste0(
      "CAST('' AS VARCHAR) AS ", quote_duckdb_identifier(column)
    ),
    character(1L)
  )
  authority <- qualified_resource_relation(
    relation = seed_catalogue_authority_relation(),
    alias = alias
  )
  clusters <- seed_catalogue_cluster_summary_cte(
    capability = capability,
    alias = alias
  )
  sequences <- seed_catalogue_sequence_cte(capability = capability, alias = alias)
  paste0(
    "WITH authority_rows AS (SELECT trim(CAST(seed_id AS VARCHAR)) AS seed_id, ",
    paste(annotations, collapse = ", "), ", ", paste(provenance, collapse = ", "),
    ", ROW_NUMBER() OVER (PARTITION BY trim(CAST(seed_id AS VARCHAR)) ORDER BY ",
    "trim(CAST(seed_id AS VARCHAR))) AS seed_row FROM ", authority,
    " WHERE nullif(trim(CAST(seed_id AS VARCHAR)), '') IS NOT NULL), authority ",
    "AS (SELECT * EXCLUDE (seed_row) FROM authority_rows WHERE seed_row = 1), ",
    clusters, ", ", sequences, " SELECT authority.seed_id, ", exact_projection,
    ", ", paste(associated_blanks, collapse = ", "), ", coalesce(clusters.",
    "source_cluster_count, 0) AS source_cluster_count, coalesce(clusters.",
    "source_cluster_ids, '') AS source_cluster_ids, coalesce(sequences.",
    "distinct_sequence_count, 0) > 0 AS sequence_available, coalesce(sequences.",
    "sequence_match_count, 0) AS sequence_match_count, coalesce(sequences.",
    "distinct_sequence_count, 0) AS distinct_sequence_count, coalesce(sequences.",
    "sequence_species, '') AS sequence_species, coalesce(sequences.",
    "sequence_identifiers, '') AS sequence_identifiers, coalesce(sequences.",
    "protein_sequence, '') AS protein_sequence, length(coalesce(sequences.",
    "protein_sequence, '')) AS protein_sequence_length, authority.source_value, ",
    "authority.source_column, authority.source_row, authority.source_path, ",
    "authority.seed_metadata_json, 'exact seed authority row' AS annotation_scope, ",
    "'known_e3_seeds' AS catalogue_source FROM authority LEFT JOIN ",
    "cluster_summary clusters USING (seed_id) LEFT JOIN seed_sequences sequences ",
    "USING (seed_id) ORDER BY lower(authority.seed_id), authority.seed_id LIMIT ",
    max_rows
  )
}

#' Build a bounded E3 seed catalogue query.
#'
#' @param capability Seed-catalogue capability metadata.
#' @param max_rows Hard row limit.
#' @param alias Attached resource alias.
#' @return DuckDB SQL query.
build_seed_catalogue_query <- function(
  capability,
  max_rows = 10000L,
  alias = "e3_resource"
) {
  if (!isTRUE(capability$available)) {
    stop("No E3 seed evidence relation is available.", call. = FALSE)
  }
  limit <- suppressWarnings(as.integer(max_rows))
  if (length(limit) != 1L || is.na(limit) || limit < 1L || limit > 100000L) {
    stop("Maximum seed rows must be between 1 and 100000.", call. = FALSE)
  }
  if (identical(capability$mode, "authority")) {
    return(build_seed_catalogue_authority_query(
      capability = capability,
      max_rows = limit,
      alias = alias
    ))
  }
  source <- qualified_resource_relation(
    relation = capability$relation,
    alias = alias
  )
  seed_column <- quote_duckdb_identifier(capability$seed_id_column)
  cluster <- if (is.null(capability$cluster_column)) {
    "CAST('' AS VARCHAR)"
  } else {
    paste0(
      "CAST(", quote_duckdb_identifier(capability$cluster_column), " AS VARCHAR)"
    )
  }
  annotations <- vapply(
    names(seed_catalogue_annotation_fields()),
    function(output) {
      seed_catalogue_aggregate(
        available = capability$columns,
        choices = seed_catalogue_annotation_fields()[[output]],
        alias = output
      )
    },
    character(1L)
  )
  sequences <- seed_catalogue_sequence_cte(capability = capability, alias = alias)
  exact_blanks <- vapply(
    seed_catalogue_exact_annotation_fields(),
    function(column) paste0(
      "CAST('' AS VARCHAR) AS ", quote_duckdb_identifier(column)
    ),
    character(1L)
  )
  paste0(
    "WITH exploded AS (SELECT DISTINCT trim(seed_id) AS seed_id, ", cluster,
    " AS source_cluster_id, * EXCLUDE (", seed_column, ") FROM ", source,
    ", UNNEST(string_split(coalesce(CAST(", seed_column,
    " AS VARCHAR), ''), ';')) AS seeds(seed_id) WHERE trim(seed_id) != ''), ",
    "seed_annotations AS (SELECT seed_id, count(DISTINCT nullif(trim(",
    "source_cluster_id), '')) AS source_cluster_count, coalesce(string_agg(",
    "DISTINCT source_cluster_id, ';' ORDER BY source_cluster_id) FILTER (WHERE ",
    "nullif(trim(source_cluster_id), '') IS NOT NULL), '') AS source_cluster_ids, ",
    paste(annotations, collapse = ", "), " FROM exploded GROUP BY seed_id), ",
    sequences, " SELECT annotations.seed_id, ",
    paste(exact_blanks, collapse = ", "), ", ",
    "annotations.associated_seed_protein_names, annotations.associated_seed_categories, ",
    "annotations.associated_seed_review_statuses, ",
    "annotations.associated_seed_ubiquitin_go_statuses, ",
    "annotations.associated_seed_organisms, annotations.source_cluster_count, ",
    "annotations.source_cluster_ids, coalesce(sequences.distinct_sequence_count, ",
    "0) > 0 AS sequence_available, coalesce(sequences.sequence_match_count, 0) ",
    "AS sequence_match_count, coalesce(sequences.distinct_sequence_count, 0) AS ",
    "distinct_sequence_count, coalesce(sequences.sequence_species, '') AS ",
    "sequence_species, coalesce(sequences.sequence_identifiers, '') AS ",
    "sequence_identifiers, coalesce(sequences.protein_sequence, '') AS ",
    "protein_sequence, length(coalesce(sequences.protein_sequence, '')) AS ",
    "protein_sequence_length, CAST('' AS VARCHAR) AS source_value, ",
    "CAST('' AS VARCHAR) AS source_column, CAST('' AS VARCHAR) AS source_row, ",
    "CAST('' AS VARCHAR) AS source_path, CAST('' AS VARCHAR) AS ",
    "seed_metadata_json, 'cluster-associated annotation; exact per-seed ",
    "linkage retained only where published by the source' AS annotation_scope, '",
    escape_sql_literal(capability$relation), "' AS catalogue_source FROM ",
    "seed_annotations annotations LEFT JOIN seed_sequences sequences USING ",
    "(seed_id) ORDER BY lower(annotations.seed_id), annotations.seed_id LIMIT ",
    limit
  )
}

#' Filter a seed catalogue using pasted literal terms.
#'
#' @param data Seed catalogue.
#' @param query One or several terms separated by common delimiters.
#' @return Filtered catalogue.
filter_seed_catalogue <- function(data, query = "") {
  terms <- unlist(strsplit(as.character(query %||% ""), "[\\n\\r\\t,;]+"))
  terms <- unique(tolower(trimws(terms[nzchar(trimws(terms))])))
  if (length(terms) == 0L || nrow(data) == 0L) return(data)
  columns <- setdiff(names(data), "protein_sequence")
  combined <- apply(data[, columns, drop = FALSE], 1L, function(row) {
    paste(ifelse(is.na(row), "", as.character(row)), collapse = " ")
  })
  keep <- vapply(combined, function(value) {
    any(vapply(terms, grepl, logical(1L), x = tolower(value), fixed = TRUE))
  }, logical(1L))
  data[keep, , drop = FALSE]
}

File: e3_python_app/src/e3app/seed_catalogue.py
"""Known-E3 seed catalogue reconstructed from the loaded resource."""

from __future__ import annotations

import logging
import re
from typing import Sequence

import duckdb
import pandas as pd

from e3app.data import list_relations, quote_identifier, relation_columns
from e3app.errors import AppError

LOGGER = logging.getLogger(__name__)

SEED_AUTHORITY_RELATION = "known_e3_seeds"
SEED_SUMMARY_RELATION_PREFERENCE = (
    "candidate_evidence",
    "e3_cluster_candidate_evidence",
)
SEED_ID_COLUMN_PREFERENCE = (
    "matched_seed_ids_calculated",
    "discovery_matched_seed_ids_calculated",
    "known_e3_seed_ids",
    "matched_e3_seeds",
)
SEED_CLUSTER_COLUMN_PREFERENCE = (
    "cluster_id",
    "representative_id",
)
SEED_SEQUENCE_RELATION = "candidate_group_member_sequences"
SEED_ANNOTATION_FIELDS = (
    (
        "associated_seed_protein_names",
        ("seed_protein_names", "discovery_seed_protein_names"),
    ),
    (
        "associated_seed_categories",
        ("seed_categories", "discovery_seed_categories"),
    ),
    (
        "associated_seed_review_statuses",
        ("seed_review_statuses", "discovery_seed_review_statuses"),
    ),
    (
        "associated_seed_ubiquitin_go_statuses",
        (
            "seed_ubiquitin_go_statuses",
            "discovery_seed_ubiquitin_go_statuses",
        ),
    ),
    (
        "associated_seed_organisms",
        ("seed_organisms", "discovery_seed_organisms"),
    ),
)
SEED_EXACT_ANNOTATION_FIELDS = (
    "seed_protein_names",
    "seed_category",
    "seed_review_status",
    "seed_ubiquitin_go_status",
    "seed_exclusion_go_term",
    "seed_organism",
    "seed_taxon_id",
    "seed_sequence_md5",
    "seed_evidence_type",
    "seed_source",
)


def _first_available(
    *,
    available: set[str],
    choices: Sequence[str],
) -> str | None:
    """Return the first available field from an ordered preference list."""
    return next((choice for choice in choices if choice in available), None)


def seed_catalogue_capability(*, connection: object) -> dict[str, object]:
    """Resolve seed-annotation and optional sequence sources.

    Args:
        connection: Open read-only DuckDB connection.

    Returns:
        Availability and source-column metadata.
    """
    relations = set(list_relations(connection))
    summary_relation: str | None = None
    summary_columns: set[str] = set()
    summary_seed_column: str | None = None
    summary_cluster_column: str | None = None
    for relation in SEED_SUMMARY_RELATION_PREFERENCE:
        if relation not in relations:
            continue
        columns = set(relation_columns(connection, relation))
        seed_id_column = _first_available(
            available=columns,
            choices=SEED_ID_COLUMN_PREFERENCE,
        )
        if seed_id_column is None:
            continue
        cluster_column = _first_available(
            available=columns,
            choices=SEED_CLUSTER_COLUMN_PREFERENCE,
        )
        summary_relation = relation
        summary_columns = columns
        summary_seed_column = seed_id_column
        summary_cluster_column = cluster_column
        break
    sequence_columns: set[str] = set()
    if SEED_SEQUENCE_RELATION in relations:
        sequence_columns = set(
            relation_columns(connection, SEED_SEQUENCE_RELATION)
        )
    sequence_available = {
        "raw_identifier",
        "protein_sequence",
    }.issubset(sequence_columns)
    if SEED_AUTHORITY_RELATION in relations:
        authority_columns = set(
            relation_columns(connection, SEED_AUTHORITY_RELATION)
        )
        if "seed_id" in authority_columns:
            return {
                "available": True,
                "mode": "authority",
                "relation": SEED_AUTHORITY_RELATION,
                "columns": tuple(sorted(authority_columns)),
                "seed_id_column": "seed_id",
                "cluster_column": None,
                "summary_relation": summary_relation,
                "summary_columns": tuple(sorted(summary_columns)),
                "summary_seed_column": summary_seed_column,
                "summary_cluster_column": summary_cluster_column,
                "sequence_available": sequence_available,
                "sequence_columns": tuple(sorted(sequence_columns)),
            }
    if summary_relation is not None:
        return {
            "available": True,
            "mode": "cluster_summary",
            "relation": summary_relation,
            "columns": tuple(sorted(summary_columns)),
            "seed_id_column": summary_seed_column,
            "cluster_column": summary_cluster_column,
            "summary_relation": summary_relation,
            "summary_columns": tuple(sorted(summary_columns)),
            "summary_seed_column": summary_seed_column,
            "summary_cluster_column": summary_cluster_column,
            "sequence_available": sequence_available,
            "sequence_columns": tuple(sorted(sequence_columns)),
        }
    return {
        "available": False,
        "mode": None,
        "relation": None,
        "columns": (),
        "seed_id_column": None,
        "cluster_column": None,
        "summary_relation": None,
        "summary_columns": (),
        "summary_seed_column": None,
        "summary_cluster_column": None,
        "sequence_available": False,
        "sequence_columns": (),
    }


def _aggregate_optional_text(
    *,
    available: set[str],
    choices: Sequence[str],
    alias: str,
) -> str:
    """Return one distinct text aggregation or a typed blank field."""
    column = _first_available(available=available, choices=choices)
    if column is None:
        return f"CAST('' AS VARCHAR) AS {quote_identifier(alias)}"
    expression = f"nullif(trim(CAST({quote_identifier(column)} AS VARCHAR)), '')"
    return (
        f"coalesce(string_agg(DISTINCT {expression}, ';' "
        f"ORDER BY {expression}) FILTER (WHERE {expression} IS NOT NULL), '') "
        f"AS {quote_identifier(alias)}"
    )


def _sequence_cte(*, capability: dict[str, object]) -> str:
    """Return optional exact seed-sequence aggregation SQL."""
    if not capability["sequence_available"]:
        return (
            "seed_sequences AS (SELECT CAST(NULL AS VARCHAR) AS seed_id, "
            "CAST(NULL AS BIGINT) AS sequence_match_count, "
            "CAST(NULL AS BIGINT) AS distinct_sequence_count, "
            "CAST(NULL AS VARCHAR) AS sequence_species, "
            "CAST(NULL AS VARCHAR) AS sequence_identifiers, "
            "CAST(NULL AS VARCHAR) AS protein_sequence WHERE FALSE)"
        )
    columns = set(capability["sequence_columns"])
    accession = (
        "CAST(parsed_accession AS VARCHAR)"
        if "parsed_accession" in columns
        else "CAST(NULL AS VARCHAR)"
    )
    species = (
        "CAST(species AS VARCHAR)"
        if "species" in columns
        else "CAST('' AS VARCHAR)"
    )
    return (
        "sequence_rows AS (SELECT coalesce(nullif(trim("
        f"{accession}), ''), nullif(trim(CAST(raw_identifier AS VARCHAR)), '')) "
        f"AS seed_id, {species} AS species, CAST(raw_identifier AS VARCHAR) "
        "AS raw_identifier, nullif(trim(CAST(protein_sequence AS VARCHAR)), '') "
        f"AS protein_sequence FROM {quote_identifier(SEED_SEQUENCE_RELATION)} "
        "), seed_sequences AS (SELECT seed_id, count(*) "
        "AS sequence_match_count, count(DISTINCT protein_sequence) FILTER "
        "(WHERE protein_sequence IS NOT NULL) AS distinct_sequence_count, "
        "coalesce(string_agg(DISTINCT species, ';' ORDER BY species) FILTER "
        "(WHERE nullif(trim(species), '') IS NOT NULL), '') AS sequence_species, "
        "coalesce(string_agg(DISTINCT raw_identifier, ';' ORDER BY raw_identifier) "
        "FILTER (WHERE nullif(trim(raw_identifier), '') IS NOT NULL), '') "
        "AS sequence_identifiers, min(protein_sequence) FILTER "
        "(WHERE protein_sequence IS NOT NULL) AS protein_sequence FROM "
        "sequence_rows WHERE seed_id IS NOT NULL GROUP BY seed_id)"
    )


def _json_text_expression(*, keys: Sequence[str], available: set[str]) -> str:
    """Return a null-safe exact metadata value from known-seed JSON."""
    if "seed_metadata_json" not in available:
        return "CAST('' AS VARCHAR)"
    candidates = [
        "nullif(trim(json_extract_string("
        "TRY_CAST(seed_metadata_json AS JSON), "
        f"'$.{key}')), '')"
        for key in keys
    ]
    return "coalesce(" + ", ".join(candidates) + ", '')"


def _authority_source_expression(*, column: str, available: set[str]) -> str:
    """Return one optional exact authority/provenance field."""
    if column not in available:
        return "CAST('' AS VARCHAR)"
    return f"coalesce(CAST({quote_identifier(column)} AS VARCHAR), '')"


def _cluster_summary_cte(*, capability: dict[str, object]) -> str:
    """Return exact seed-to-cluster links where cluster summaries exist."""
    relation = capability.get("summary_relation")
    seed_column = capability.get("summary_seed_column")
    cluster_column = capability.get("summary_cluster_column")
    if relation is None or seed_column is None or cluster_column is None:
        return (
            "cluster_summary AS (SELECT CAST(NULL AS VARCHAR) AS seed_id, "
            "CAST(NULL AS BIGINT) AS source_cluster_count, "
            "CAST(NULL AS VARCHAR) AS source_cluster_ids WHERE FALSE)"
        )
    return (
        "cluster_links AS (SELECT DISTINCT trim(seed_id) AS seed_id, "
        f"CAST({quote_identifier(str(cluster_column))} AS VARCHAR) "
        f"AS source_cluster_id FROM {quote_identifier(str(relation))}, "
        "UNNEST(string_split(coalesce(CAST("
        f"{quote_identifier(str(seed_column))} AS VARCHAR), ''), ';')) "
        "AS seeds(seed_id) WHERE trim(seed_id) != ''), "
        "cluster_summary AS (SELECT seed_id, count(DISTINCT "
        "nullif(trim(source_cluster_id), '')) AS source_cluster_count, "
        "coalesce(string_agg(DISTINCT source_cluster_id, ';' ORDER BY "
        "source_cluster_id) FILTER (WHERE nullif(trim(source_cluster_id), '') "
        "IS NOT NULL), '') AS source_cluster_ids FROM cluster_links "
        "GROUP BY seed_id)"
    )


def _build_authority_seed_catalogue_query(
    *,
    capability: dict[str, object],
    maximum_rows: int,
) -> str:
    """Build a catalogue from the exact normalised known-seed authority."""
    available = set(capability["columns"])
    metadata = {
        "seed_protein_names": ("protein_names", "protein_name", "name"),
        "seed_category": ("category", "e3_category"),
        "seed_review_status": ("reviewed", "review_status"),
        "seed_ubiquitin_go_status": ("ubiquitin_go_term",),
        "seed_exclusion_go_term": ("exclusion_go_term",),
        "seed_organism": ("organism",),
        "seed_taxon_id": ("organism_id", "taxon_id"),
        "seed_sequence_md5": ("sequence_md5",),
        "seed_evidence_type": ("evidence_type",),
        "seed_source": ("source",),
    }
    annotations = ", ".join(
        f"{_json_text_expression(keys=keys, available=available)} "
        f"AS {quote_identifier(alias)}"
        for alias, keys in metadata.items()
    )
    provenance = ", ".join(
        f"{_authority_source_expression(column=column, available=available)} "
        f"AS {quote_identifier(column)}"
        for column in (
            "source_value",
            "source_column",
            "source_row",
            "source_path",
            "seed_metadata_json",
        )
    )
    cluster_summary = _cluster_summary_cte(capability=capability)
    sequence_cte = _sequence_cte(capability=capability)
    blank_associations = ", ".join(
        f"CAST('' AS VARCHAR) AS {quote_identifier(alias)}"
        for alias, _ in SEED_ANNOTATION_FIELDS
    )
    exact_projection = ", ".join(
        f"authority.{quote_identifier(column)}"
        for column in SEED_EXACT_ANNOTATION_FIELDS
    )
    return f"""
        WITH authority_rows AS (
            SELECT trim(CAST(seed_id AS VARCHAR)) AS seed_id,
                   {annotations},
                   {provenance},
                   ROW_NUMBER() OVER (
                       PARTITION BY trim(CAST(seed_id AS VARCHAR))
                       ORDER BY trim(CAST(seed_id AS VARCHAR))
                   ) AS seed_row
            FROM {quote_identifier(SEED_AUTHORITY_RELATION)}
            WHERE nullif(trim(CAST(seed_id AS VARCHAR)), '') IS NOT NULL
        ), authority AS (
            SELECT * EXCLUDE (seed_row) FROM authority_rows WHERE seed_row = 1
        ), {cluster_summary}, {sequence_cte}
        SELECT authority.seed_id,
               {exact_projection},
               {blank_associations},
               coalesce(clusters.source_cluster_count, 0)
                   AS source_cluster_count,
               coalesce(clusters.source_cluster_ids, '') AS source_cluster_ids,
               coalesce(sequences.distinct_sequence_count, 0) > 0
                   AS sequence_available,
               coalesce(sequences.sequence_match_count, 0)
                   AS sequence_match_count,
               coalesce(sequences.distinct_sequence_count, 0)
                   AS distinct_sequence_count,
               coalesce(sequences.sequence_species, '') AS sequence_species,
               coalesce(sequences.sequence_identifiers, '')
                   AS sequence_identifiers,
               coalesce(sequences.protein_sequence, '') AS protein_sequence,
               length(coalesce(sequences.protein_sequence, ''))
                   AS protein_sequence_length,
               authority.source_value,
               authority.source_column,
               authority.source_row,
               authority.source_path,
               authority.seed_metadata_json,
               'exact seed authority row' AS annotation_scope,
               '{SEED_AUTHORITY_RELATION}' AS catalogue_source
        FROM authority
        LEFT JOIN cluster_summary clusters USING (seed_id)
        LEFT JOIN seed_sequences sequences USING (seed_id)
        ORDER BY lower(authority.seed_id), authority.seed_id
        LIMIT {int(maximum_rows)}
    """


def build_seed_catalogue_query(
    *,
    capability: dict[str, object],
    maximum_rows: int,
) -> str:
    """Build a bounded seed catalogue query.

    Args:
        capability: Metadata returned by :func:`seed_catalogue_capability`.
        maximum_rows: Hard row limit from 1 to 100,000.

    Returns:
        Executable DuckDB SQL.

    Raises:
        AppError: If source metadata or the row bound is invalid.
    """
    if not capability.get("available"):
        raise AppError("No E3 seed evidence relation is available")
    if not 1 <= maximum_rows <= 100_000:
        raise AppError("maximum seed rows must be between 1 and 100000")
    if capability.get("mode") == "authority":
        return _build_authority_seed_catalogue_query(
            capability=capability,
            maximum_rows=maximum_rows,
        )
    relation = str(capability["relation"])
    available = set(capability["columns"])
    seed_column = str(capability["seed_id_column"])
    cluster_column = capability["cluster_column"]
    cluster_expression = (
        f"CAST({quote_identifier(str(cluster_column))} AS VARCHAR)"
        if cluster_column is not None
        else "CAST('' AS VARCHAR)"
    )
    annotation_sql = ", ".join(
        _aggregate_optional_text(
            available=available,
            choices=choices,
            alias=alias,
        )
        for alias, choices in SEED_ANNOTATION_FIELDS
    )
    sequence_cte = _sequence_cte(capability=capability)
    blank_exact = ", ".join(
        f"CAST('' AS VARCHAR) AS {quote_identifier(column)}"
        for column in SEED_EXACT_ANNOTATION_FIELDS
    )
    return f"""
        WITH exploded AS (
            SELECT DISTINCT
                   trim(seed_id) AS seed_id,
                   {cluster_expression} AS source_cluster_id,
                   * EXCLUDE ({quote_identifier(seed_column)})
            FROM {quote_identifier(relation)},
                 UNNEST(string_split(
                     coalesce(CAST({quote_identifier(seed_column)} AS VARCHAR), ''),
                     ';'
                 )) AS seeds(seed_id)
            WHERE trim(seed_id) != ''
        ), seed_annotations AS (
            SELECT seed_id,
                   count(DISTINCT nullif(trim(source_cluster_id), ''))
                       AS source_cluster_count,
                   coalesce(string_agg(DISTINCT source_cluster_id, ';'
                       ORDER BY source_cluster_id) FILTER (
                           WHERE nullif(trim(source_cluster_id), '') IS NOT NULL
                       ), '') AS source_cluster_ids,
                   {annotation_sql}
            FROM exploded
            GROUP BY seed_id
        ), {sequence_cte}
        SELECT annotations.seed_id,
               {blank_exact},
               annotations.associated_seed_protein_names,
               annotations.associated_seed_categories,
               annotations.associated_seed_review_statuses,
               annotations.associated_seed_ubiquitin_go_statuses,
               annotations.associated_seed_organisms,
               annotations.source_cluster_count,
               annotations.source_cluster_ids,
               coalesce(sequences.distinct_sequence_count, 0) > 0
                   AS sequence_available,
               coalesce(sequences.sequence_match_count, 0)
                   AS sequence_match_count,
               coalesce(sequences.distinct_sequence_count, 0)
                   AS distinct_sequence_count,
               coalesce(sequences.sequence_species, '') AS sequence_species,
               coalesce(sequences.sequence_identifiers, '')
                   AS sequence_identifiers,
               coalesce(sequences.protein_sequence, '') AS protein_sequence,
               length(coalesce(sequences.protein_sequence, ''))
                   AS protein_sequence_length,
               CAST('' AS VARCHAR) AS source_value,
               CAST('' AS VARCHAR) AS source_column,
               CAST('' AS VARCHAR) AS source_row,
               CAST('' AS VARCHAR) AS source_path,
               CAST('' AS VARCHAR) AS seed_metadata_json,
               'cluster-associated annotation; exact per-seed linkage retained '
               'only where published by the source' AS annotation_scope,
               '{relation}' AS catalogue_source
        FROM seed_annotations annotations
        LEFT JOIN seed_sequences sequences USING (seed_id)
        ORDER BY lower(annotations.seed_id), annotations.seed_id
        LIMIT {int(maximum_rows)}
    """


def collect_seed_catalogue(
    *,
    connection: duckdb.DuckDBPyConnection,
    maximum_rows: int = 10_000,
) -> pd.DataFrame:
    """Collect the bounded seed catalogue and available exact sequences."""
    capability = seed_catalogue_capability(connection=connection)
    query = build_seed_catalogue_query(
        capability=capability,
        maximum_rows=maximum_rows,
    )
    try:
        result = connection.execute(query).fetchdf()
    except duckdb.Error as exc:
        LOGGER.exception("Could not collect the E3 seed catalogue")
        raise AppError(f"Could not collect the E3 seed catalogue: {exc}") from exc
    LOGGER.info("Collected %s E3 seed catalogue rows", len(result))
    return result


def filter_seed_catalogue(*, frame: pd.DataFrame, query: str) -> pd.DataFrame:
    """Filter seeds using one or several literal pasted terms."""
    terms = tuple(
        dict.fromkeys(
            token.strip().casefold()
            for token in re.split(r"[\n\r\t,;]+", str(query or ""))
            if token.strip()
        )
    )
    if not terms or frame.empty:
        return frame
    searchable = [
        column
        for column in frame.columns
        if column != "protein_sequence"
    ]
    combined = frame[searchable].fillna("").astype(str).agg(" ".join, axis=1)
    folded = combined.str.casefold()
    mask = pd.Series(False, index=frame.index)
    for term in terms:
        mask |= folded.str.contains(term, regex=False)
    return frame.loc[mask].copy()
